File: csharp/EviFluor/kits.cs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

using System;
using System.Text.Json.Nodes;
using Hse.EviFluor;

namespace Hse.EviFluor.Kits;

/// <summary>
/// Selects the mathematical model used to convert the interpolated signal into a concentration.
/// </summary>
public enum FitAlgorithm
{
    /// <summary>
    /// Uses a linear function <c>k1 * x + k2</c>.
    /// </summary>
    Linear = 1,
    /// <summary>
    /// Uses a power-law function <c>k1 * x^k2</c>.
    /// </summary>
    Power = 2,
    /// <summary>
    /// Uses a quadratic function <c>k1 * x^2 + k2 * x + k3</c>.
    /// </summary>
    Quadratic = 3,
}

/// <summary>
/// Generic configurable kit implementation used directly or as base class for predefined kits.
/// </summary>
public class Default : IKit
{
    private const string KitFitAlgorithm = "fitAlgorithm";
    private const string KitK1 = "k1";
    private const string KitK2 = "k2";
    private const string KitK3 = "k3";
    private const string KitSettlingTime = "settlingTime";
    private const string KitStdHighTargetSignalFactor = "stdHighTargetSignalFactor";
    private const string KitDescription = "description";

    private readonly FitAlgorithm FitAlgorithm_;
    private readonly double K1_;
    private readonly double K2_;
    private readonly double K3_;
    private readonly double SettlingTime_;
    private readonly double? StdHighTargetSignalFactor_;
    private readonly string Description_;

    /// <summary>
    /// Initializes a kit with explicit fit parameters and optional measurement behavior.
    /// </summary>
    /// <param name="fitAlgorithm">Fit algorithm applied after the standard-based interpolation step.</param>
    /// <param name="k1">Primary fit parameter.</param>
    /// <param name="k2">Secondary fit parameter.</param>
    /// <param name="k3">Third fit parameter used by the quadratic model.</param>
    /// <param name="settlingTime">Delay in seconds before a sample measurement is acquired.</param>
    /// <param name="stdHighTargetSignalFactor">Optional target factor used during first-sample auto-gain.</param>
    /// <param name="description">Human-readable kit description.</param>
    public Default(
        FitAlgorithm fitAlgorithm = FitAlgorithm.Linear,
        double k1 = 1.0,
        double k2 = 0.0,
        double k3 = 0.0,
        double settlingTime = 0.0,
        double? stdHighTargetSignalFactor = null,
        string? description = null)
    {
        FitAlgorithm_ = fitAlgorithm;
        K1_ = k1;
        K2_ = k2;
        K3_ = k3;
        SettlingTime_ = settlingTime;
        StdHighTargetSignalFactor_ = stdHighTargetSignalFactor;
        Description_ = description ?? "Default kit with linear fit";
    }

    /// <summary>
    /// Returns a diagnostic string containing the kit description and parameters.
    /// </summary>
    /// <returns>Text representation of the configured kit.</returns>
    public override string ToString()
    {
        return $"{Description_} fit_algorithm={FitAlgorithm_}, k1={K1_}, k2={K2_}, k3={K3_}, settling_time={SettlingTime_}, std_high_target_signal_factor={StdHighTargetSignalFactor_}";
    }

    /// <summary>
    /// Serializes the kit configuration to JSON.
    /// </summary>
    /// <returns>JSON node containing the configurable kit fields.</returns>
    public JsonNode ToJson()
    {
        JsonObject obj = new JsonObject
        {
            [KitFitAlgorithm] = JsonValue.Create((int)FitAlgorithm_),
            [KitK1] = JsonValue.Create(K1_),
            [KitK2] = JsonValue.Create(K2_),
            [KitK3] = JsonValue.Create(K3_),
            [KitSettlingTime] = JsonValue.Create(SettlingTime_),
            [KitDescription] = JsonValue.Create(Description_),
        };

        obj[KitStdHighTargetSignalFactor] = StdHighTargetSignalFactor_ is null
            ? null
            : JsonValue.Create(StdHighTargetSignalFactor_.Value);
        return obj;
    }

    /// <summary>
    /// Deserializes a configurable kit from JSON created by <see cref="ToJson"/>.
    /// </summary>
    /// <param name="node">JSON node containing the kit fields.</param>
    /// <returns>Deserialized <see cref="Default"/> instance.</returns>
    public static Default FromJson(JsonNode node)
    {
        JsonObject obj = node as JsonObject ?? throw new InvalidOperationException("Kit JSON must be an object");
        int fitAlgorithmValue = obj[KitFitAlgorithm]?.GetValue<int>()
            ?? throw new InvalidOperationException($"{KitFitAlgorithm} is missing or null");

        if (!Enum.IsDefined(typeof(FitAlgorithm), fitAlgorithmValue))
        {
            throw new InvalidOperationException($"Unknown fit algorithm: {fitAlgorithmValue}");
        }

        return new Default(
            fitAlgorithm: (FitAlgorithm)fitAlgorithmValue,
            k1: obj[KitK1]?.GetValue<double>() ?? throw new InvalidOperationException($"{KitK1} is missing or null"),
            k2: obj[KitK2]?.GetValue<double>() ?? throw new InvalidOperationException($"{KitK2} is missing or null"),
            k3: obj[KitK3]?.GetValue<double>() ?? 0.0,
            settlingTime: obj[KitSettlingTime]?.GetValue<double>() ?? 0.0,
            stdHighTargetSignalFactor: obj[KitStdHighTargetSignalFactor]?.GetValue<double>(),
            description: obj[KitDescription]?.GetValue<string>());
    }

    /// <summary>
    /// Creates a predefined kit instance from a string name.
    /// </summary>
    /// <param name="name">Case-insensitive kit name.</param>
    /// <returns>Matching predefined kit instance.</returns>
    /// <exception cref="ArgumentException">Thrown when the name does not match a known kit.</exception>
    public static Default Factory(string name)
    {
        string normalized = name.ToLowerInvariant();

        if (normalized == "default")
        {
            return new Default();
        }
        if (normalized == "qubittm_1x_dsdna_broad_range_br" || normalized == "qubit_br")
        {
            return new QubitTM_1X_dsDNA_Broad_Range_BR();
        }
        if (normalized == "qubittm_1x_dsdna_high_sensitivity_hs" || normalized == "qubit_hs")
        {
            return new QubitTM_1X_dsDNA_High_Sensitivity_HS();
        }

        throw new ArgumentException($"Unknown kit: {name}", nameof(name));
    }

    /// <summary>
    /// Calculates the concentration for a measured signal value using the configured fit model.
    /// </summary>
    /// <param name="stdLow">Measured low standard.</param>
    /// <param name="stdHigh">Measured high standard.</param>
    /// <param name="value">Measured sample signal.</param>
    /// <returns>Calculated concentration.</returns>
    public double fit(Point stdLow, Point stdHigh, double value)
    {
        double deltaSignal = stdHigh.Value - stdLow.Value;
        if (deltaSignal == 0)
        {
            throw new InvalidOperationException("Standard low and high must have different signal values for calibration");
        }

        var m = (stdHigh.Concentration - stdLow.Concentration) / deltaSignal;
        var b = stdHigh.Concentration - m * stdHigh.Value;
        var x = m * value + b;

        return FitAlgorithm_ switch
        {
            FitAlgorithm.Linear => K1_ * x + K2_,
            FitAlgorithm.Power => x < 0 ? 0.0 : K1_ * Math.Pow(x, K2_),
            FitAlgorithm.Quadratic => x < 0 ? 0.0 : K1_ * Math.Pow(x, 2) + K2_ * x + K3_,
            _ => throw new InvalidOperationException($"Unknown fit algorithm: {FitAlgorithm_}"),
        };
    }

    /// <summary>
    /// Returns the default settling time for the kit in seconds.
    /// </summary>
    /// <returns>Configured settling time in seconds.</returns>
    public double SettlingTime()
    {
        return SettlingTime_;
    }

    /// <summary>
    /// Returns the optional target signal factor used during first-sample auto-gain.
    /// </summary>
    /// <returns>Configured factor, or <c>null</c> when no override is defined.</returns>
    public double? StdHighTargetSignalFactor()
    {
        return StdHighTargetSignalFactor_;
    }
}

/// <summary>
/// Predefined kit for the Qubit 1X dsDNA high-sensitivity assay.
/// </summary>
public class QubitTM_1X_dsDNA_High_Sensitivity_HS : Default
{
    /// <summary>
    /// Initializes the Qubit 1X dsDNA high-sensitivity preset.
    /// </summary>
    public QubitTM_1X_dsDNA_High_Sensitivity_HS()
        : base(
            fitAlgorithm: FitAlgorithm.Linear,
            k1: 1.0,
            k2: 0.0,
            description: "QubitTM 1X dsDNA High Sensitivity (HS) Assay Kit",
            settlingTime: 0.0)
    {
    }
}

/// <summary>
/// Predefined kit for the Qubit 1X dsDNA broad-range assay.
/// </summary>
public class QubitTM_1X_dsDNA_Broad_Range_BR : Default
{
    /// <summary>
    /// Initializes the Qubit 1X dsDNA broad-range preset.
    /// </summary>
    public QubitTM_1X_dsDNA_Broad_Range_BR()
        : base(
            fitAlgorithm: FitAlgorithm.Power,
            k1: 1.011658,
            k2: 1.014932,
            description: "QubitTM 1X dsDNA Broad Range (BR) Assay Kit",
            stdHighTargetSignalFactor: 0.4,
            settlingTime: 0.0)
    {
    }
}

File: python/hse/evifluor/kits.py
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

from enum import IntEnum


_KIT_FIT_ALGORITHM = "fitAlgorithm"
_KIT_K1 = "k1"
_KIT_K2 = "k2"
_KIT_K3 = "k3"
_KIT_SETTLING_TIME = "settlingTime"
_KIT_STD_HIGH_TARGET_SIGNAL_FACTOR = "stdHighTargetSignalFactor"
_KIT_DESCRIPTION = "description"


class FitAlgorithm(IntEnum):
    """Supported post-calibration fit curves for kit-specific concentration mapping."""
    Linear                                        = 1
    Power                                         = 2
    Quadratic                                     = 3
    HillFit                                       = 4

class Default:
    """Base kit configuration combining standard interpolation with an optional fit curve."""

    def __init__(self, fit_algorithm = FitAlgorithm.Linear, k1 = 1.0, k2 = 0.0, k3 = 0.0, settling_time = 0.0, std_high_target_signal_factor = None, description = None):
        try:
            self._fit_algorithm = FitAlgorithm(fit_algorithm)
        except ValueError as exc:
            raise ValueError("Unknown fit algorithm: {}".format(fit_algorithm)) from exc
        self._k1 = k1
        self._k2 = k2
        self._k3 = k3
        self._settling_time = settling_time
        self._std_high_target_signal_factor = std_high_target_signal_factor
        self._description = description or "Default kit with linear fit"
        
    def __repr__(self):
        """Returns a textual description of the kit."""
        return f"{self._description} fit_algorithm={self._fit_algorithm.name}, k1={self._k1}, k2={self._k2}, k3={self._k3}, settling_time={self._settling_time}, std_high_target_signal_factor={self._std_high_target_signal_factor}"

    def to_json(self):
        """Converts the kit configuration to a JSON representation."""
        return {
            _KIT_FIT_ALGORITHM: int(self._fit_algorithm),
            _KIT_K1: self._k1,
            _KIT_K2: self._k2,
            _KIT_K3: self._k3,
            _KIT_SETTLING_TIME: self._settling_time,
            _KIT_STD_HIGH_TARGET_SIGNAL_FACTOR: self._std_high_target_signal_factor,
            _KIT_DESCRIPTION: self._description,
        }

    @staticmethod
    def factory(name):
        """Creates a predefined kit instance from a string name."""
        name = name.lower()

        if name == "default":
            return Default()
        if name == "qubittm_1x_dsdna_broad_range_br" or name == "qubit_br":
            return QubitTM_1X_dsDNA_Broad_Range_BR()
        if name == "qubittm_1x_dsdna_high_sensitivity_hs" or name == "qubit_hs":
            return QubitTM_1X_dsDNA_High_Sensitivity_HS()

        raise ValueError("Unknown kit: {}".format(name))

    @staticmethod
    def from_json(node):
        """Creates a kit instance from a JSON node."""
        return Default(
            fit_algorithm = node[_KIT_FIT_ALGORITHM],
            k1 = node[_KIT_K1],
            k2 = node[_KIT_K2],
            k3 = node.get(_KIT_K3, 0.0),
            settling_time = node.get(_KIT_SETTLING_TIME, 0.0),
            std_high_target_signal_factor = node.get(_KIT_STD_HIGH_TARGET_SIGNAL_FACTOR),
            description = node.get(_KIT_DESCRIPTION),
        )

    def fit(self, std_low, std_high, rfu):
        """Maps a measured signal to concentration using the calibrated standards and fit curve."""
        # First interpolate between the two measured standards to get the raw concentration.
        delta_signal = std_high.value - std_low.value
        if delta_signal == 0:
            raise ValueError("Standard low and high must have different signal values for calibration")

        m = (std_high.concentration - std_low.concentration) / delta_signal
        b = std_high.concentration - m * std_high.value
        x = m * rfu + b
        
        # Then apply the kit-specific correction curve on top of the raw interpolation result.
        if self._fit_algorithm == FitAlgorithm.Linear:
            x_fitted = self._k1 * x + self._k2
        elif self._fit_algorithm == FitAlgorithm.Power:
            # Negative concentrations are clamped to zero before exponentiation.
            if x < 0:
                x_fitted = 0
            else:
                x_fitted = self._k1 * x ** self._k2
        elif self._fit_algorithm == FitAlgorithm.Quadratic:
            # Negative concentrations are clamped to zero before the quadratic correction.
            if x < 0:
                x_fitted = 0
            else:
                x_fitted = self._k1 * x ** 2 + self._k2 * x + self._k3
        elif self._fit_algorithm == FitAlgorithm.HillFit:
            # Negative concentrations are clamped to zero before the correction.
            # k1 : VMAX
            # k2 : K
            # k3 : n
            
            num = std_high.value * (std_high.concentration - std_low.concentration)
            den = (std_high.concentration - std_low.concentration) ** 2

            if den == 0:
                raise ValueError("All concentrations are zero")

            S = num / den
            
            if rfu < 0:
                x_fitted = 0
            else:
                x = rfu / S
                x_fitted = (self._k1 * x ** self._k3) / (self._k2 ** self._k3 + x ** self._k3)
        return x_fitted
    
    def settling_time(self):
        """Returns the time in seconds to wait after adding the reagent before measuring."""
        return self._settling_time

    def std_high_target_signal_factor(self):
        """Returns the autogain factor for the high standard signal."""
        return self._std_high_target_signal_factor

class QubitTM_1X_dsDNA_High_Sensitivity_HS(Default):
    """Preset for the QubitTM 1X dsDNA High Sensitivity (HS) Assay Kit."""

    def __init__(self):
        """Initializes the QubitTM 1X dsDNA High Sensitivity (HS) Assay Kit preset."""
        super().__init__(fit_algorithm = FitAlgorithm.Linear, k1 = 1.0, k2 = 0.0, description = "QubitTM 1X dsDNA High Sensitivity (HS) Assay Kit", settling_time = 0.0)
        pass

class QubitTM_1X_dsDNA_Broad_Range_BR(Default):
    """Preset for the QubitTM 1X dsDNA Broad Range (BR) Assay Kit."""

    def __init__(self):
        super().__init__(fit_algorithm = FitAlgorithm.HillFit, k1 = 52452044.021155454, k2 = 3342216.695097067 , k3 = 1.2623036398450653, description = "QubitTM 1X dsDNA Broad Range (BR) Assay Kit", std_high_target_signal_factor = 0.4, settling_time = 0.0)
        pass
